stew/result.nim

```diff
@@ -151,40 +173,42 @@ type
       v: T
 
 func raiseResultError[T, E](self: Result[T, E]) {.noreturn.} =
+  mixin toException
+
   when E is ref Exception:
     if self.e.isNil: # for example Result.default()!
-      raise ResultError[void](msg: "Trying to access value with err (nil)")
+      raise (ref ResultError[void])(msg: "Trying to access value with err (nil)")
     raise self.e
-  elif compiles(self.e.toException()):
-    raise self.e.toException()
+  elif compiles(toException(self.e)):
+    raise toException(self.e)
   elif compiles($self.e):
-    raise ResultError[E](
+    raise (ref ResultError[E])(
       error: self.e, msg: "Trying to access value with err: " & $self.e)
   else:
-    raise ResultError[E](error: self.e)
+    raise (res ResultError[E])(msg: "Trying to access value with err", error: self.e)
 
-template ok*(R: type Result, x: auto): auto =
+template ok*[T, E](R: type Result[T, E], x: auto): R =
   ## Initialize a result with a success and value
   ## Example: `Result[int, string].ok(42)`
   R(o: true, v: x)
 
-template ok*(self: var Result, x: auto) =
+template ok*[T, E](self: var Result[T, E], x: auto) =
   ## Set the result to success and update value
   ## Example: `result.ok(42)`
   self = ok(type self, x)
 
-template err*(R: type Result, x: auto): auto =
+template err*[T, E](R: type Result[T, E], x: auto): R =
```

Comment:
Believe it or not, but the short version of the code will be actually more robust (the generic param binding in templates is currently broken) and it will compile slightly faster on every call-site :)

Reply:
Believe it or not, but the nim compiler has bugs: https://github.com/nim-lang/Nim/issues/13799

Reply:
(yes, devel too)